### DISCUSSION TITLE
Why does ui.context_menu() not work for ui.tree?

### DISCUSSION THREAD

**Discussion by cwfbss** [zauberzeug/nicegui#4010] [Q&A] [answered]
### Question

Wy does the ui.context_menu() not work for ui.tree ? 
It would be the easiest way to trigger any action on nodes or leaefs.

**Comment by falkoschindler** [comment]
@cwfbss Can you, please, try to create a minimum reproducible code example? What have you tried? What exactly does not work? This would allow us and the community to help more efficiently. Thanks!
[Why is it important to provide a minimal reproducible example?](https://github.com/zauberzeug/nicegui/wiki/FAQs#why-is-it-important-to-provide-a-minimal-reproducible-example)

**Comment by cwfbss** [comment]
just for the sake of simplicity:

```py
with ui.tree([
    {'id': 'numbers', 'children': [{'id': '1'}, {'id': '2'}]},
    {'id': 'letters', 'children': [{'id': 'A'}, {'id': 'B'}]},
], label_key='id', on_select=lambda e: ui.notify(e.value)):
     with ui.context_menu():
        ui.menu_item('..any action on current selection..')
        ui.menu_item('Move directory / file to...')
        ui.separator()
        ui.menu_item('Reset', auto_close=False)
```

**Reply by falkoschindler** [reply, marked as answer]
I see. The problem seems to be that `with ui.tree(): ...` want's to populate the tree's default slot, but there is no such slot. Quasar's QTree only has slots for individual node headers and bodies. You can use them to add context menus like this:

```py
tree = ui.tree([
    {'id': 'numbers', 'children': [{'id': '1'}, {'id': '2'}]},
    {'id': 'letters', 'children': [{'id': 'A'}, {'id': 'B'}]},
], label_key='id')
tree.add_slot('default-header', '''
    <span :props="props">Node <strong>{{ props.node.id }}</strong>
        <q-menu touch-position context-menu>
            <q-item clickable v-close-popup @click="() => props.tree.$emit('click_open', props.key)">
                <q-item-section>Open</q-item-section>
            </q-item>
        </q-menu>
    </span>
''')
tree.on('click_open', lambda e: ui.notify(f'Open "{e.args}"'))
```

**Reply by cwfbss** [reply]
Ok, thank You ! I see, eventually it comes down to the implementation of Quasar's QTree and Your suggested solution help me to carry on.
